File: JavaScript/968-Binary Tree Cameras.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

// 0: leaf node or node not covered
// 1: with camera, covered
// 2: without camera, covered

var minCameraCover = function(root) {
//Sol 1
    let res = 0;
    let helper = (node) =>{
        if (!node) return 2;
        let left = helper(node.left), right = helper(node.right);
        if (left === 0 || right === 0){
            res += 1;
            return 1;
        }else if ( left === 1 || right === 1){
            return 2;
        }else{
            return 0;
        }
    }
    return (helper(root) === 0)?res+1:res; 
    
// Sol 2
    let covered = new Set();
    covered.add(null);
    
    let dfs = (node, parent) =>{
        if (!node) return 0;
        let result = dfs(node.left, node) + dfs(node.right, node);
        if (!covered.has(node.left) || !covered.has(node.right)){
            covered.add(node.left);covered.add(node.right);covered.add(node);covered.add(parent);
            result += 1;
        }
        return result;
    }
    res = dfs(root, null);
    
    return covered.has(root)?res:res+1;
};

File: C++/968-Binary Tree Cameras.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<pair<TreeNode *, int>, int> memo;  // use map, but this requires O(logN) time to look up
    int cal(TreeNode* root, int state){
        if (!root) return 0;
        auto key = make_pair(root, state);
        if (memo.find(key) != memo.end()) return memo[key];
        
        int res;
        if (state == 0){
            res = 1 + cal(root->left, 1) + cal(root->right, 1);
            if (root->left) res = min(res, 1+cal(root->left, 2) + cal(root->right, 0) );
            if (root->right) res = min(res, 1+cal(root->right, 2) + cal(root->left, 0) );
        } else if (state == 1){
            res = min(1+cal(root->left, 1)+cal(root->right, 1), cal(root->left, 0)+cal(root->right, 0) );
        } else{
            res = cal(root->left, 1) + cal(root->right, 1);
        }
        memo[key] = res;
        return res;
    }
    
    int minCameraCover(TreeNode* root) {
        // define states, 0: not covered, 1: covered and not occupied, 2: covered and occupied
        return cal(root, 0);
    }
};